Shared/Realm.Sync.Shared/Handles/SharedRealmHandleExtensions.cs

```diff
@@ -153,11 +171,32 @@ private static unsafe void RefreshAccessTokenCallback(IntPtr userHandlePtr, IntP
         #if __IOS__
         [ObjCRuntime.MonoPInvokeCallback(typeof(NativeMethods.SessionErrorCallback))]
         #endif
-        private static unsafe void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorCode, sbyte* messageBuffer, IntPtr messageLength)
+        private static unsafe void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorCode, sbyte* messageBuffer, IntPtr messageLength, IntPtr userInfoPairs, int userInfoPairsLength)
         {
             var session = Session.Create(sessionHandlePtr);
             var message = new string(messageBuffer, 0, (int)messageLength, System.Text.Encoding.UTF8);
-            Session.RaiseError(session, new SessionErrorException(message, errorCode));
+
+            SessionException exception;
+
+            switch (errorCode)
+            {
+                case ErrorCode.DivergingHistories:
+                    var userInfo = MarshalErrorUserInfo(userInfoPairs, userInfoPairsLength);
+                    exception = new ClientResetException(message, userInfo);
+                    break;
+                default:
+                    exception = new SessionException(message, errorCode);
+                    break;
+            }
+
+            Session.RaiseError(session, exception);
+        }
+
+        private static Dictionary<string, string> MarshalErrorUserInfo(IntPtr userInfoPairs, int userInfoPairsLength)
+        {
+            return Enumerable.Range(0, userInfoPairsLength)
+                             .Select(i => Marshal.PtrToStructure<KeyValuePair<string, string>>(IntPtr.Add(userInfoPairs, i * kvpStringStringSize)))
```

Comment:
Will this marshal correctly convert from C++ char* ANSI null-terminated strings?